src/snowflake/snowpark/profiler.py

```diff
@@ -0,0 +1,210 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+import re
+from contextlib import contextmanager
+from typing import List, Optional
+
+import snowflake.snowpark
+from snowflake.snowpark._internal.utils import validate_object_name
+
+
+class Profiler:
+    """
+    Setup profiler to receive profiles of stored procedures.
+
+    Note:
+        This feature cannot be used in owner's right SP because owner's right SP will not be able to set session-level parameters.
+    """
+
+    def __init__(
+        self,
+        stage: Optional[str] = "",
+        active_profiler: Optional[str] = "LINE",
+        session: Optional["snowflake.snowpark.Session"] = None,
+    ) -> None:
+        self.stage = stage
+        self.active_profiler = active_profiler
+        self.modules_to_register = []
+        self.register_modules_sql = ""
+        self.set_targeted_stage_sql = ""
+        self.enable_profiler_sql = ""
+        self.disable_profiler_sql = ""
+        self.set_active_profiler_sql = ""
+        self.pattern = r"WITH\s+.*?\s+AS\s+PROCEDURE\s+.*?\s+CALL\s+.*"
+        self.session = session
+        self._prepare_sql()
+        self.query_history = None
```

Comment:
do we want those member variable to be public? if not we can add "_"  (self._stage) in front of them to make them private